### PR TITLE
ghostGhostedBoundaries

### PR DESCRIPTION
**Pull request by fdkong**
Setup periodic boundary conditions in a scalable way  when the mesh is generated in parallel

Refs #15501

### CODE REVIEW

**Comment by lindsayad** [comment]
> Refs #00000

That's a lot of zeros

**Comment by fdkong** [comment]
> > Refs #00000
> 
> That's a lot of zeros

I was wondering if civet has an overflow issue as well by using a lot of zeros

**Comment by fdkong** [comment]
We need it for a geometric search.

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -1569,6 +1571,19 @@ MooseMesh::detectPairedSidesets()
...
+  // collect information from everyone.
+  if (_use_distributed_mesh)
+    for (unsigned side_dim = 0; side_dim < dim; ++side_dim)
+    {
+      _communicator.set_union(minus_x_ids[side_dim]);
+      _communicator.set_union(plus_x_ids[side_dim]);
+      _communicator.set_union(minus_y_ids[side_dim]);
+      _communicator.set_union(plus_y_ids[side_dim]);
+      _communicator.set_union(minus_z_ids[side_dim]);
+      _communicator.set_union(plus_z_ids[side_dim]);
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:1584]: You're the parallel guy so I'll defer to you on this, but is 6 communications in a row ideal? I'm wondering if there's a way to do this all in one go, but if you don't expect this to be important than I'm not going to worry about it.

File: framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C [hunk 2]
```diff
@@ -1015,6 +1023,10 @@ DistributedRectilinearMeshGenerator::buildCube(UnstructuredMesh & mesh,
   bool old_allow_renumbering = mesh.allow_renumbering();
   mesh.allow_renumbering(false);
   mesh.allow_find_neighbors(false);
+  // No, I do not want to remove anything in case periodic boundary conditions
+  // are required late. We  stop libMesh from deleting the valuable remote
+  // elements paired with the local elements
+  mesh.allow_remote_element_removal(false);
```
Comment (lindsayad) [inline framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C:1029]: @roystgnr doesn't the default libmesh geometric ghosting functor consider periodic boundaries? Perhaps we're just not telling libmesh about periodic boundaries soon enough? Again I would like to get to a point where we can feel comfortable deleting remote elements at any point during mesh construction; to get to that point we need to be confident that we've added the geometric ghosting functors that we need

File: test/include/userobjects/CheckGhostedBoundaries.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "GeneralUserObject.h"
+
+/**
+ * A native way to check if all boundaries are gathered to every single processor.
```
Comment (lindsayad) [inline test/include/userobjects/CheckGhostedBoundaries.h:15]: Do you mean "native" or do you mean "naive"?

File: test/src/userobjects/CheckGhostedBoundaries.C [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+void
+CheckGhostedBoundaries::execute()
+{
+  dof_id_type nelems = _fe_problem.mesh().getMesh().get_boundary_info().build_side_list().size();
+
+  // If the total number of boundary sides is the same as the number of local boundary sides,
+  // we will conclude all-gather happens. It is not necessarily true, but it works fine
+  // for setting up tests
+  if (_total_num_bdry_sides == nelems)
+    mooseError("All boundaries are ghosted to every single processor, and it is not scalable.");
```
Comment (lindsayad) [inline test/src/userobjects/CheckGhostedBoundaries.C:36]: not scalable = not beautiful :smile:

File: test/tests/bcs/dmg_periodic/tests [hunk 5]
```diff
@@ -0,0 +1,58 @@
+[Tests]
+  issues = '#15501'
+  design = 'AddPeriodicBCAction.md  DistributedRectilinearMeshGenerator.md'
+
+  [./check]
+    type = 'RunException'
+    input = 'dmg_periodic_bc.i'
+    cli_args = '/UserObjects/uo/type=CheckGhostedBoundaries /UserObjects/uo/total_num_bdry_sides=160'
+    expect_err = 'All boundaries are ghosted to every single processor, and it is not scalable'
+    requirement = "MOOSE shall check ghosted boundaries"
```
Comment (lindsayad) [inline test/tests/bcs/dmg_periodic/tests:10]: Should this be something more like:
"The system shall be able to determine if a non-scalable algorithm is being used for ghosting boundaries."
?

Comment (lindsayad) [reply test/tests/bcs/dmg_periodic/tests:10]: You could ask @permcody to be sure, but I think we prefer "The system" over "MOOSE".

File: test/tests/bcs/dmg_periodic/tests [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  [../]
+
+  [./check_one_step]
+    type = 'Exodiff'
+    input = 'dmg_periodic_bc.i'
+    exodiff = 'dmg_periodic_bc_out_one_step.e'
+    cli_args = '/UserObjects/uo/type=CheckGhostedBoundaries /UserObjects/uo/total_num_bdry_sides=160 Outputs/hide="pid" Outputs/file_base=dmg_periodic_bc_out_one_step Executioner/num_steps=1'
+    abs_zero = 1e-6
+    rel_err = 1e-5
+    requirement = "MOOSE shall check ghosted boundaries for one time step"
```
Comment (lindsayad) [inline test/tests/bcs/dmg_periodic/tests:21]: This requirement is unclear to me. What's really the goal of this test? We shall check and conclude what?

File: test/tests/bcs/dmg_periodic/tests [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  [./2d]
+    type = 'Exodiff'
+    input = 'dmg_periodic_bc.i'
+    exodiff = 'dmg_periodic_bc_out.e'
+    min_parallel = 2
+    cli_args = 'Outputs/hide="pid"'
+    abs_zero = 1e-6
+    rel_err = 1e-5
+    requirement = "MOOSE shall support periodic boundary conditions in a parallel generated mesh."
```
Comment (lindsayad) [inline test/tests/bcs/dmg_periodic/tests:33]: This requirement is clear to me :+1:

Comment (lindsayad) [reply test/tests/bcs/dmg_periodic/tests:33]: Maybe explicitly say that this is for 2D. I'm just guessing that's the dimension since the next two requirements are for 3D and 1D

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/mesh/MooseMesh.C (hunk 1 above)
Comment (fdkong) [reply framework/src/mesh/MooseMesh.C:1584]: Actually more than 6. There are 6*dim communications. I did not find an API that can do all these using one global communication. 

@roystgnr Any API? I want to do  6*dim communications all at once.

File: framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C (hunk 2 above)
Comment (fdkong) [reply framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C:1029]: I agreed totally.

 I would like to give minimal information, and `libmesh` takes care of others. The biggest issue is that `libmesh` is able to delete elements, but it can not add elements back. That being said,  I would like to see `libmesh` being able to grow the local subdomain if needed. 

If we can increase the subdomain automatically, then I do not even need to add ghosting elements in a mesh generator.

Based on the current capability we have, the choice I have right now is to stop `libmesh` from deleting anything we added. 

Periodic boundary setup needs to know the mesh information... the whole process is a chicken-egg story.

File: test/include/userobjects/CheckGhostedBoundaries.h (hunk 3 above)
Comment (fdkong) [reply test/include/userobjects/CheckGhostedBoundaries.h:15]: Yes.

Comment (fdkong) [reply test/include/userobjects/CheckGhostedBoundaries.h:15]: I will change ..

File: test/tests/bcs/dmg_periodic/tests (hunk 5 above)
Comment (fdkong) [reply test/tests/bcs/dmg_periodic/tests:10]: Sometimes we use moose, and sometimes we use system .....

File: test/tests/bcs/dmg_periodic/tests (hunk 6 above)
Comment (fdkong) [reply test/tests/bcs/dmg_periodic/tests:21]: I conclude that boundaries are actually distributed.

File: test/tests/bcs/dmg_periodic/tests (hunk 7 above)
Comment (fdkong) [reply test/tests/bcs/dmg_periodic/tests:33]: Yes, it is 2D

File: test/src/userobjects/CheckGhostedBoundaries.C (hunk 4 above)
Comment (fdkong) [reply test/src/userobjects/CheckGhostedBoundaries.C:36]: Yeah.

File: test/tests/bcs/dmg_periodic/tests (hunk 6 above)
Comment (lindsayad) [reply test/tests/bcs/dmg_periodic/tests:21]: Let's write that into the requirement

File: test/tests/bcs/dmg_periodic/tests (hunk 5 above)
Comment (fdkong) [reply test/tests/bcs/dmg_periodic/tests:10]: The system seems to be more general.

File: framework/src/mesh/MooseMesh.C (hunk 1 above)
Comment (roystgnr) [reply framework/src/mesh/MooseMesh.C:1584]: I'm afraid we don't have an API to do that without repacking into a different data structure first.  In this case a vector of tuples of 6 boundary_id_type would work best if your sets are nearly the same size, or a vector of ```pair<boundary_id_type, int>``` would work better (where the int is an index from 0 to 5 identifying which set each came from) if the vector of tuples would be too sparse.

The only other alternative is something I've been thinking about for a while - adding optional Request arguments to a lot more APIs, so you can interleave computation and communication even on things like gathers and reductions.  That'll be more long term, but if you're interested then I'll bump it up from "idle speculation" to "thing I actually intend to do eventually".  ;-)

File: framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C (hunk 2 above)
Comment (roystgnr) [reply framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C:1029]: The PeriodicBoundary just needs to know a translation vector or function, and then you just need a boundary_id_type to tell it where it applies.  What mesh information do you need to know?

But yeah - the default ghosting considers periodic boundaries, but you have to have the PeriodicBoundary specifications all attached before you ever reach a prepare_for_use() with allow_remote_element_removal() still set to true.  Setting it to false temporarily in the case where you're already doing distributed generation is probably the best workaround here.

However ... you're already doing distributed generation - you're also manually ghosting potential periodic neighbors, then, I hope?  Relying on MeshCommunication::gather to pull them in wouldn't work, IIRC.

Comment (fdkong) [reply framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C:1029]: Yes, for this special mesh generator `DistributedRectilinearMeshGenerator`, I would like to keep anything I generate, regardless. So I would like to just leave my code as is.

Thanks, everyone!

**Comment by fdkong** [comment]
It is ready for review or merge again @lindsayad

File: test/tests/bcs/dmg_periodic/tests [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  issues = '#15501'
+  design = 'AddPeriodicBCAction.md  DistributedRectilinearMeshGenerator.md'
+
+  [./check]
+    type = 'RunException'
+    input = 'dmg_periodic_bc.i'
+    cli_args = '/UserObjects/uo/type=CheckGhostedBoundaries /UserObjects/uo/total_num_bdry_sides=160'
+    expect_err = 'All boundaries are ghosted to every single processor, and it is not scalable'
+    requirement = "The system shall be able to determine if a non-scalable algorithm is being used for ghosting boundaries."
+    max_parallel = 1
```
Comment (lindsayad) [inline test/tests/bcs/dmg_periodic/tests:11]: This error only works in serial?

Comment (fdkong) [reply test/tests/bcs/dmg_periodic/tests:11]: Yes, I intended to add this test to make sure nobody can delete my code: `CheckGhostedBoundaries` class.

Error is triggered when boundaries are NOT distributed. If there is one core, then boundaries are not distributed in the sense that the local number of side elements equals to the global number of side elements.

Comment (fdkong) [reply test/tests/bcs/dmg_periodic/tests:11]: Boundaries in the mesh generated via `DistributedRectilinearMeshGenerator` are always distributed.  So only way to trigger that little error is to use one core.

**Review by lindsayad — APPROVED** [review]
beautiful code
